### PR TITLE
Implement transaction hash

### LINKED ISSUE #458: Transaction implementation
The Transaction module should do the following

- compute the transaction hash for a given transaction

### PR DESCRIPTION
**Pull request by anshulforyou**
Fixes | Closes | Resolves #
Closes #458 

Description
Added the transaction hash logic for all types of transactions supported on starknet.

Changes:
Added declare transaction hash logic
Added deploy transaction hash logic
Added invoke transaction hash logic

Types of changes
New feature (non-breaking change which adds functionality)

Testing
Requires testing: Yes

Did you write tests??: Yes

Documentation
If this requires a documentation update, did you add one? No

### CODE REVIEW

**Comment by jelilat** [comment]
Please close https://github.com/NethermindEth/juno/pull/479 in favour of this

File: core/transaction.go [hunk 1]
```diff
@@ -26,10 +31,40 @@ type DeployTransaction struct {
 	Version *felt.Felt
 }
 
-func (d *DeployTransaction) Hash() *felt.Felt {
-	// Todo: implement pedersen hash as defined here:
+func (d *DeployTransaction) Hash(chainId []byte) (*felt.Felt, error) {
```
Comment (joshklop) [inline core/transaction.go:34]: We can shorten this function by doing something like this:
```go
func (d *DeployTransaction) Hash(chainId *felt.Felt) (*felt.Felt, error) {
	snKeccakContructor, err := crypto.StarkNetKeccak([]byte("constructor"))
	if err != nil {
		return nil, err
	}
	return crypto.PedersenArray(
		new(felt.Felt).SetBytes([]byte("deploy")),
		d.Version,
		d.ContractAddress,
		snKeccakContructor,
		crypto.PedersenArray(d.ConstructorCalldata...),
		new(felt.Felt),
		chainId,
	), nil
}
```

File: core/transaction.go [hunk 2]
```diff
@@ -57,15 +92,64 @@ type InvokeTransaction struct {
 	Version *felt.Felt
 }
 
-func (i *InvokeTransaction) Hash() *felt.Felt {
+func (i *InvokeTransaction) Hash(chainId []byte) (*felt.Felt, error) {
```
Comment (joshklop) [inline core/transaction.go:95]: We can shorten this function by doing something like this:
```go
func (it *InvokeTransaction) Hash(chainId *felt.Felt) (*felt.Felt, error) {
	invokeFelt := new(felt.Felt).SetBytes([]byte("invoke"))
	if it.Version.IsZero() {
		return crypto.PedersenArray(
			invokeFelt,
			it.Version,
			it.ContractAddress,
			it.EntryPointSelector,
			crypto.PedersenArray(it.CallData...),
			it.MaxFee,
			chainId,
		), nil
	} else if it.Version.IsOne() {
		return crypto.PedersenArray(
			invokeFelt,
			it.Version,
			it.SenderAddress,
			new(felt.Felt),
			crypto.PedersenArray(it.CallData...),
			it.MaxFee,
			chainId,
			it.Nonce,
		), nil
	}
	return nil, errors.New("invalid transaction version")
}
```

**Review by joshklop — CHANGES_REQUESTED** [review]
Nice work so far. Can you please add tests for version zero of the deploy transaction and version zero of the invoke transaction? That is also a good opportunity to refactor the existing tests to follow [table-driven testing](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) and other standard go testing practices.

File: core/transaction.go (hunk 1 above)
Comment (anshulforyou) [reply core/transaction.go:34]: Done

File: core/transaction.go (hunk 2 above)
Comment (anshulforyou) [reply core/transaction.go:77]: Done

**Comment by IronGauntlets** [comment]
@anshulforyou can you please add the links which you used to get the test data?

**Review by IronGauntlets — APPROVED** [review]

**Review by joshklop — APPROVED** [review]
